source/http2_stream_manager.c

```diff
@@ -390,11 +394,119 @@ static void s_aws_http2_stream_manager_build_transaction_synced(struct aws_http2
+        "PINGACK received for connection: %p. Schedule another PING to be sent.",
+        (void *)sm_connection->connection);
+    sm_connection->thread_data.ping_received = true;
+}
+
+static void s_connection_ping_timeout_task(struct aws_channel_task *task, void *arg, enum aws_task_status status) {
+    (void)task;
+    (void)status;
+    struct aws_h2_sm_connection *sm_connection = arg;
+    if (status != AWS_TASK_STATUS_RUN_READY) {
+        goto done;
+    }
+    if (!sm_connection->connection) {
+        /* The connection has been released before timeout happens, just release the refcount */
+        goto done;
+    }
+    AWS_ASSERT(aws_channel_thread_is_callers_thread(aws_http_connection_get_channel(sm_connection->connection)));
+    if (!sm_connection->thread_data.ping_received) {
+        /* Timeout happened */
+        STREAM_MANAGER_LOGF(
+            ERROR,
+            sm_connection->stream_manager,
+            "ping timeout detected for connection: %p, closing connection.",
+            (void *)sm_connection->connection);
+
+        aws_http_connection_close(sm_connection->connection);
+    } else {
+        struct aws_channel *channel = aws_http_connection_get_channel(sm_connection->connection);
+        /* acquire a refcount for next set of tasks to run */
+        aws_ref_count_acquire(&sm_connection->ref_count);
+        aws_channel_schedule_task_future(
+            channel, &sm_connection->ping_task, sm_connection->thread_data.next_ping_task_time);
+    }
+done:
+    /* Release refcount for current set of tasks */
+    aws_ref_count_release(&sm_connection->ref_count);
+}
+
+static void s_connection_ping_task(struct aws_channel_task *task, void *arg, enum aws_task_status status) {
+    (void)task;
+    (void)status;
+    struct aws_h2_sm_connection *sm_connection = arg;
+    if (status != AWS_TASK_STATUS_RUN_READY) {
+        aws_ref_count_release(&sm_connection->ref_count);
+        return;
+    }
+    if (!sm_connection->connection) {
+        /* The connection has been released before ping task, just release the refcount */
+        aws_ref_count_release(&sm_connection->ref_count);
+        return;
+    }
+    AWS_ASSERT(aws_channel_thread_is_callers_thread(aws_http_connection_get_channel(sm_connection->connection)));
+
+    STREAM_MANAGER_LOGF(
+        TRACE, sm_connection->stream_manager, "Sending PING for connection: %p.", (void *)sm_connection->connection);
+    aws_http2_connection_ping(sm_connection->connection, NULL, s_on_ping_complete, sm_connection);
+    sm_connection->thread_data.ping_received = false;
+    struct aws_channel *channel = aws_http_connection_get_channel(sm_connection->connection);
+    uint64_t schedule_time = 0;
+    aws_channel_current_clock_time(channel, &schedule_time);
+    sm_connection->thread_data.next_ping_task_time =
+        schedule_time + sm_connection->stream_manager->connection_ping_period_ns;
+    schedule_time += sm_connection->stream_manager->connection_ping_timeout_ns;
```

Comment:
trivial: the `schedule_time` variable is a bit confusing ... where it's used as "current time" on line 488 but then we update it to be "timeout time" on line 489.

Can we just have two different variables: rename ~schedule_times~ -> `current_time` and then have a new variable later like `uint64_t timeout_time = current_time + sm_connection->stream_manager->connection_ping_timeout_ns`